### PR TITLE
Proper implementation of CROSS APPLY that fixes behavior of CROSS JOIN on parenthesized table expressions

### PR DESCRIPTION
**Pull request by Sairakan**
### Description

The previous attempt to fix CROSS JOIN ended up causing several regressions with CROSS APPLY. Move
CROSS/OUTER APPLY parsing back into the bison parser, and this time resolve the issues with CROSS JOIN
by adding new TSQL_CROSS and TSQL_OUTER tokens that are reserved keywords to remove the ambiguity that
previously existed where CROSS and OUTER_P could be interpreted as function calls.
 
### Issues Resolved

BABEL-328
BABEL-2805

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-decl.y [hunk 1]
```diff
@@ -104,10 +104,13 @@
 	TSQL_MUST_CHANGE TSQL_CHECK_EXPIRATION TSQL_CHECK_POLICY TSQL_CREDENTIAL TSQL_SID TSQL_OLD_PASSWORD
 	TSQL_UNLOCK TSQL_VALUES
 	TSQL_NVARCHAR
+    TSQL_CROSS TSQL_OUTER TSQL_APPLY
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-decl.y]: Could you use the same indentation as other code? Indentations are different in most of the files in this PR.

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y [hunk 2]
```diff
@@ -1511,8 +1511,82 @@ table_ref:	relation_expr tsql_table_hint_expr
...
+                    n->alias = $3;
+                    $$ = (Node *) n;
+                }
+            | TSQL_APPLY select_with_parens opt_alias_clause
+                {
+                    RangeSubselect *n = makeNode(RangeSubselect);
+                    n->lateral = true;
+                    n->subquery = $2;
+                    n->alias = $3;
+                    /* same comment as above */
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y]: What does it mean?

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y [hunk 3]
```diff
@@ -1511,8 +1511,82 @@ table_ref:	relation_expr tsql_table_hint_expr
...
+                                        errmsg("VALUES in FROM must have an alias"),
+                                        errhint("For example, FROM (VALUES ...) [AS] foo."),
+                                        parser_errposition(@2)));
+                        else
+                            ereport(ERROR,
+                                    (errcode(ERRCODE_SYNTAX_ERROR),
+                                        errmsg("subquery in FROM must have an alias"),
+                                        errhint("For example, FROM (SELECT ...) [AS] foo."),
+                                        parser_errposition(@2)));
+                    }
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y]: Do we have tests for those error cases?

File: test/JDBC/expected/BABEL-328.out [hunk 4]
```diff
@@ -126,6 +126,14 @@ RETURN
 ) 
 GO
 
+CREATE PROC p AS
```
Comment (2jungkook) [inline test/JDBC/expected/BABEL-328.out]: 1. Let's rewrite the whole BABEL-328 as a form of MVU test including views and procs.
2. Please make sure every parser rules you introduced are tested.

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y (hunk 2 above)
Comment (Sairakan) [reply contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y]: sorry, this is taken from the equivalent engine rule, which is referencing this related comment:
/*
* The SQL spec does not permit a subselect
* (<derived_table>) without an alias clause,
* so we don't either.  This avoids the problem
* of needing to invent a unique refname for it.
* That could be surmounted if there's sufficient
* popular demand, but for now let's just implement
* the spec and see if anyone complains.
* However, it does seem like a good idea to emit
* an error message that's better than "syntax error".
*/

I'll put the whole comment here as well.

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y (hunk 3 above)
Comment (Sairakan) [reply contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y]: No, I will ad a few

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-decl.y (hunk 1 above)
Comment (Sairakan) [reply contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-decl.y]: Done

File: test/JDBC/expected/BABEL-328.out (hunk 4 above)
Comment (Sairakan) [reply test/JDBC/expected/BABEL-328.out]: Done

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y [hunk 5]
```diff
@@ -1511,6 +1511,83 @@ table_ref:	relation_expr tsql_table_hint_expr
...
+						* The SQL spec does not permit a subselect
+						* (<derived_table>) without an alias clause,
+						* so we don't either.  This avoids the problem
+						* of needing to invent a unique refname for it.
+						* That could be surmounted if there's sufficient
+						* popular demand, but for now let's just implement
+						* the spec and see if anyone complains.
+						* However, it does seem like a good idea to emit
+						* an error message that's better than "syntax error".
+						*/
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y]: indentation

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y [hunk 6]
```diff
@@ -1511,6 +1511,83 @@ table_ref:	relation_expr tsql_table_hint_expr
 					n->relation = (Node *) $1;
 					$$ = (Node *) n;
 				}
+			| TSQL_APPLY func_table func_alias_clause
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y:1514]: Let's add a test when func_alias_clause is used and referenced by another filter/join condition.

File: contrib/babelfishpg_tsql/src/backend_parser/parser.c [hunk 7]
```diff
@@ -167,6 +167,12 @@ pgtsql_base_yylex(YYSTYPE *lvalp, YYLTYPE *llocp, core_yyscan_t yyscanner)
 		case SERVER:
 			cur_token_length = 6;
 			break;
+        case CROSS:
+            cur_token_length = 5;
+            break;
+        case OUTER_P:
+            cur_token_length = 5;
+            break;
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/src/backend_parser/parser.c]: indentation

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y (hunk 5 above)
Comment (Sairakan) [reply contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y]: Fixed

File: contrib/babelfishpg_tsql/src/backend_parser/parser.c (hunk 7 above)
Comment (Sairakan) [reply contrib/babelfishpg_tsql/src/backend_parser/parser.c]: Fixed

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y (hunk 6 above)
Comment (Sairakan) [reply contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y:1514]: Added

**Review by 2jungkook — APPROVED** [review]

**Pull request merged**